src/SparkConduit.sol

```diff
@@ -0,0 +1,314 @@
+    }
+
+    // -- Storage --
+
+    mapping(address => AssetData) private assets;
+
+    /// @inheritdoc ISparkConduit
+    address public roles;
+    /// @inheritdoc ISparkConduit
+    address public registry;
+    /// @inheritdoc ISparkConduit
+    uint256 public subsidySpread;
+
+    // -- Immutable/constant --
+
+    /// @inheritdoc ISparkConduit
+    IPool   public immutable pool;
+    /// @inheritdoc ISparkConduit
+    address public immutable pot;
+
+    uint256 private constant RAY = 10 ** 27;
+    uint256 private constant SECONDS_PER_YEAR = 365 days;
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "SparkConduit/not-authorized");
+        _;
+    }
+
+    modifier ilkAuth(bytes32 ilk) {
+        require(
+            RolesLike(roles).canCall(ilk, msg.sender, address(this), msg.sig),
+            "SparkConduit/ilk-not-authorized"
+        );
+        _;
+    }
+
+    constructor(IPool  _pool, address _pot) {
+        pool = _pool;
+        pot  = _pot;
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function deposit(bytes32 ilk, address asset, uint256 amount) external ilkAuth(ilk) {
+        require(assets[asset].enabled, "SparkConduit/asset-disabled");
+        require(
+            assets[asset].positions[ilk].pendingWithdrawals == 0,
+            "SparkConduit/no-deposit-with-pending-withdrawals"
+        );
+        require(amount <= maxDeposit(ilk, asset), "SparkConduit/max-deposit-exceeded");
+
+        address source = RegistryLike(registry).buffers(ilk);
+
+        require(
+            IERC20(asset).transferFrom(source, address(this), amount),
+            "SparkConduit/transfer-failed"
+        );
+
+        pool.supply(asset, amount, address(this), 0);
+
+        // Convert asset amount to shares
+        uint256 shares = amount.rayDiv(pool.getReserveData(asset).liquidityIndex);
+
+        assets[asset].positions[ilk].shares += shares;
+        assets[asset].totalShares           += shares;
+
+        emit Deposit(ilk, asset, source, amount);
+    }
+
+    /// @inheritdoc IAllocatorConduit
+    function withdraw(bytes32 ilk, address asset, uint256 maxAmount)
+        external ilkAuth(ilk) returns (uint256 amount)
+    {
+        uint256 liquidityAvailable
+            = IERC20(asset).balanceOf(pool.getReserveData(asset).aTokenAddress);
+
+        amount = liquidityAvailable < maxAmount ? liquidityAvailable : maxAmount;
+
+        // Normally you should update local state first for re-entrancy,
+        // but we need an update-to-date liquidity index for that
+        amount = pool.withdraw(asset, amount, address(this));
+
+        // Convert asset amount to shares
+        uint256 liquidityIndex = pool.getReserveData(asset).liquidityIndex;
+        uint256 shares         = amount.rayDiv(liquidityIndex);
+
+        uint256 pshares     = assets[asset].positions[ilk].shares;
+        uint256 withdrawals = assets[asset].positions[ilk].pendingWithdrawals;
+
+        if (pshares < shares) {
```

Comment:
Should we not just prevent this behaviour althogether? Should we add a check before the withdraw to ensure that an ilk can't withdraw more than their aTokens corresponding underlying?

Reply:
Could do a maxWithdraw check looks like

Reply:
Is it because of the value of `liquidityIndex`?

Reply:
This has been rewritten.